src/Base/Matrix.cpp

```diff
@@ -832,37 +832,42 @@ ScaleType Matrix4D::hasScale(double tol) const
     // Only in case of uniform scaling it doesn't make a difference.
     if (tol == 0.0)
         tol = 1e-9;
+    
+    // compare absolute values propotional to size
+    auto close = [&](double a, double b) {
+        return fabs(fabs(a/b)-1.0) <= tol;
+    };
 
     // get column vectors
-    double dx = Vector3d(dMtrx4D[0][0],dMtrx4D[1][0],dMtrx4D[2][0]).Sqr();
-    double dy = Vector3d(dMtrx4D[0][1],dMtrx4D[1][1],dMtrx4D[2][1]).Sqr();
-    double dz = Vector3d(dMtrx4D[0][2],dMtrx4D[1][2],dMtrx4D[2][2]).Sqr();
+    double dx = getCol(0).Sqr();
+    double dy = getCol(1).Sqr();
+    double dz = getCol(2).Sqr();
     double dxyz = sqrt(dx * dy * dz);
 
     // get row vectors
-    double du = Vector3d(dMtrx4D[0][0],dMtrx4D[0][1],dMtrx4D[0][2]).Sqr();
-    double dv = Vector3d(dMtrx4D[1][0],dMtrx4D[1][1],dMtrx4D[1][2]).Sqr();
-    double dw = Vector3d(dMtrx4D[2][0],dMtrx4D[2][1],dMtrx4D[2][2]).Sqr();
+    double du = getRow(0).Sqr();
+    double dv = getRow(1).Sqr();
+    double dw = getRow(2).Sqr();
     double duvw = sqrt(du * dv * dw);
 
     double d3 = determinant3();
 
     // This could be e.g. a projection, a shearing,... matrix
-    if (fabs(dxyz - d3) > tol && fabs(duvw - d3) > tol) {
+    if (!close(dxyz, d3) && !close(duvw, d3)) {
         return ScaleType::Other;
     }
 
-    if (fabs(duvw - d3) <= tol && (fabs(du - dv) > tol || fabs(dv - dw) > tol)) {
+    if (close(duvw, d3) && (!close(du, dv) || !close(dv, dw))) {
         return ScaleType::NonUniformLeft;
     }
 
-    if (fabs(dxyz - d3) <= tol && (fabs(dx - dy) > tol || fabs(dy - dz) > tol)) {
+    if (close(dxyz, d3) && (!close(dx, dy) || !close(dy, dz))) {
         return ScaleType::NonUniformRight;
     }
 
-    if (fabs(dx - 1.0) > tol) {
+    if (abs(d3 - 1.0) > tol) {

```

Comment:
Be careful with fabs vs abs (or std::abs). While for fabs it's guaranteed to work as expected for floating point numbers it depends with std::abs. There are several overloaded functions implemented and if not the right header is included the floating point numbers are cast to int and thus std::abs leads to unexpected results.
In this case it works because cmath is included by Vector3D.h

In the past I have read several times that fabs should be replaced with abs but according to https://en.cppreference.com/w/cpp/numeric/math/fabs it's still a  valid function.

Reply:
Ops. I did read about that and was aware, but I did it anyway.